### PR TITLE
cli: list new remote branches during git fetch

### PR DESCRIPTION
**Pull request by 0xdeafbeef**
Adds imported branches list for git fetch.
```
jj git fetch
From origin
* [new untracked branch]  test2
 ```
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by 0xdeafbeef** [comment]
I'm not sure what to do with these [tests](https://github.com/martinvonz/jj/blob/26528091e646a79a3724997f06c4af69eaf70297/cli/tests/test_git_fetch.rs#L229-L260). They are designed to expect no stdout in the event of an error. I can address this issue, but I'm unsure if I should.

**Comment by yuja** [comment]
> I'm not sure what to do with these [tests](https://github.com/martinvonz/jj/blob/26528091e646a79a3724997f06c4af69eaf70297/cli/tests/test_git_fetch.rs#L229-L260). They are designed to expect no stdout in the event of an error.

Perhaps, the output can be written to stderr. It's kind of a status message.

**Comment by 0xdeafbeef** [comment]
> > I'm not sure what to do with these [tests](https://github.com/martinvonz/jj/blob/26528091e646a79a3724997f06c4af69eaf70297/cli/tests/test_git_fetch.rs#L229-L260). They are designed to expect no stdout in the event of an error.
> 
> Perhaps, the output can be written to stderr. It's kind of a status message.

I hadn't noticed that 'Abandoned commits' were already being written to stderr. With these changes, all tests now pass.

File: lib/src/git.rs [hunk 1]
```diff
@@ -201,10 +201,12 @@ impl GitImportError {
 }
 
 /// Describes changes made by `import_refs()` or `fetch()`.
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Eq, PartialEq, Default)]
 pub struct GitImportStats {
     /// Commits superseded by newly imported commits.
     pub abandoned_commits: Vec<CommitId>,
+    /// Newly fetched remote refs.
+    pub new_remote_refs: BTreeSet<RefName>,
```
Comment (yuja) [inline lib/src/git.rs:209]: nit: rename to `changed_remote_refs`? "new" might be confusing because the set contains created/deleted/moved refs.

File: lib/src/git.rs [hunk 2]
```diff
@@ -326,7 +328,10 @@ pub fn import_some_refs(
     } else {
         vec![]
     };
-    let stats = GitImportStats { abandoned_commits };
+    let stats = GitImportStats {
+        abandoned_commits,
+        new_remote_refs: BTreeSet::from_iter(changed_remote_refs.keys().cloned()),
```
Comment (yuja) [inline lib/src/git.rs:333]: simply return the `changed_remote_refs` map? The caller would want to know if the ref is created `(absent, present)` or deleted `(present, absent)`.

File: cli/src/git_util.rs [hunk 3]
```diff
@@ -172,7 +172,45 @@ pub fn with_remote_git_callbacks<T>(
...
+        for refname in &stats.new_remote_refs {
+            match refname {
+                RefName::RemoteBranch { branch, remote } => {
+                    new_remote_branches.entry(remote).or_default().push(branch);
+                }
+                RefName::Tag(tag) => tags.push(tag),
+                RefName::LocalBranch(b) => local_branches.push(b),
+            }
+        }
+        local_branches.sort_unstable();
```
Comment (yuja) [inline cli/src/git_util.rs:195]: nit: no need to sort the list again? The original set is sorted by `RefName`.

FWIW, I personally don't mind if branches and tags are displayed in one big list, something like:
```
branch: main@origin (tracked)
branch: feature1@origin (deleted)
..
tag: 1.2.3 (created)
```

It will simplify the post processing code.

File: cli/src/git_util.rs [hunk 4]
```diff
@@ -172,7 +172,45 @@ pub fn with_remote_git_callbacks<T>(
...
+        for branch in local_branches {
+            writeln!(ui.stderr(), "* [imported local branch]  {branch}")?;
+        }
+        for (remote, branches) in new_remote_branches.iter_mut() {
+            writeln!(ui.stderr(), "From {remote}")?;
+            branches.sort_unstable();
+            for branch in branches {
+                writeln!(ui.stderr(), "* [new untracked branch]  {branch}")?;
```
Comment (yuja) [inline cli/src/git_util.rs:203]: "new" and "untracked" are wrong.

The ref might be just moved or deleted. To check if it's moved/created/deleted, `(old_remote_ref, new_target)` pair can be used.

To check if it's untracked/tracked, we'll probably need to query `repo.view().get_remote_branch(branch, remote).is_tracking()`.

File: cli/src/git_util.rs [hunk 5]
```diff
@@ -172,7 +172,45 @@ pub fn with_remote_git_callbacks<T>(
     f(callbacks)
 }
 
-pub fn print_git_import_stats(ui: &mut Ui, stats: &GitImportStats) -> Result<(), CommandError> {
+pub fn print_git_import_stats(
+    ui: &mut Ui,
+    stats: &GitImportStats,
+    show_ref_stats: bool,
+) -> Result<(), CommandError> {
+    use std::collections::BTreeMap;
```
Comment (yuja) [inline cli/src/git_util.rs:180]: nit: move the `use` statement to top

File: cli/src/git_util.rs [hunk 6]
```diff
@@ -172,7 +172,45 @@ pub fn with_remote_git_callbacks<T>(
...
+                RefName::RemoteBranch { branch, remote } => {
+                    new_remote_branches.entry(remote).or_default().push(branch);
+                }
+                RefName::Tag(tag) => tags.push(tag),
+                RefName::LocalBranch(b) => local_branches.push(b),
+            }
+        }
+        local_branches.sort_unstable();
+        for branch in local_branches {
+            writeln!(ui.stderr(), "* [imported local branch]  {branch}")?;
```
Comment (yuja) [inline cli/src/git_util.rs:197]: nit: optionally the branch name can be colorized by using the formatter API. (see `cmd_branch_list()`.)

It doesn't have to be implemented now, so probably better to do in a separate PR.

File: cli/src/commands/git.rs [hunk 7]
```diff
@@ -751,7 +751,7 @@ fn do_git_clone(
             unreachable!("we didn't provide any globs")
         }
     })?;
-    print_git_import_stats(ui, &stats.import_stats)?;
+    print_git_import_stats(ui, &stats.import_stats, false)?;
```
Comment (yuja) [inline cli/src/commands/git.rs:754]: nit: might be better to print the imported refs on `jj git clone`/`import`, but that can be decided later.

**Review by yuja — COMMENTED** [review]
Thanks!

File: lib/src/git.rs (hunk 1 above)
Comment (0xdeafbeef) [reply lib/src/git.rs]: resolved

File: lib/src/git.rs (hunk 2 above)
Comment (0xdeafbeef) [reply lib/src/git.rs]: resolved

File: cli/src/git_util.rs (hunk 3 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs]: Done

File: cli/src/git_util.rs (hunk 4 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs]: Partially resolved, see
https://github.com/martinvonz/jj/pull/3044/files#diff-9bc1bb41a6477444bc8efccd5a0c7b053997c4aa67a216e87f8ee819bedbc7acR239-R245

Comment (0xdeafbeef) [reply cli/src/git_util.rs]: Resolved

File: cli/src/git_util.rs (hunk 6 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs]: I'll do it in a separate pr

File: cli/src/commands/git.rs (hunk 7 above)
Comment (0xdeafbeef) [reply cli/src/commands/git.rs]: I wasn't sure what to do with local branches, so skipped it for now. Can turn it on if you think it makes sense

File: cli/src/git_util.rs [hunk 8]
```diff
@@ -172,17 +173,122 @@ pub fn with_remote_git_callbacks<T>(
     f(callbacks)
 }
 
-pub fn print_git_import_stats(ui: &mut Ui, stats: &GitImportStats) -> Result<(), CommandError> {
+pub fn print_git_import_stats(
+    ui: &mut Ui,
+    stats: &GitImportStats,
+    query_ref_stats: Option<&ReadonlyRepo>,
+) -> Result<(), CommandError> {
```
Comment (yuja) [inline cli/src/git_util.rs:180]: nit: better to pass `repo` and `show_ref_stats` separately because there may be other things that requires `repo`.

`(Ui, dyn Repo, GitImportStats, bool)`

File: cli/src/git_util.rs [hunk 9]
```diff
@@ -172,17 +173,122 @@ pub fn with_remote_git_callbacks<T>(
...
+        for (refname, (remote_ref, ref_target)) in &stats.changed_remote_refs {
+            let ref_status = RefStatus::new(refname, remote_ref, ref_target, repo);
+            match refname {
+                RefName::RemoteBranch { .. } => remote_branches.push(ref_status),
+                RefName::Tag(_) => tags.push(ref_status),
+                RefName::LocalBranch(_) => local_branches.push(ref_status),
+            }
+        }
+
+        let grouped_refs = vec![remote_branches, tags, local_branches];
```
Comment (yuja) [inline cli/src/git_util.rs:196]: nit: no need to split the changes into groups since the list is sorted anyway.

File: cli/src/git_util.rs [hunk 10]
```diff
@@ -172,17 +173,122 @@ pub fn with_remote_git_callbacks<T>(
...
+
+        for refs in grouped_refs {
+            if let Some(max_ref_name_len) = refs.iter().map(|status| status.ref_name.len()).max() {
```
Comment (yuja) [inline cli/src/git_util.rs:199]: to calculate the padding width, you can use `UnicodeWidth::width`. See `show_diff_stat()` for the usage example.

We can of course implement the padding later as a separate patch.

File: cli/src/git_util.rs [hunk 11]
```diff
@@ -172,17 +173,122 @@ pub fn with_remote_git_callbacks<T>(
...
+                format!("branch: {branch}@{remote}")
+            }
+            RefName::Tag(tag) => format!("tag: {tag}"),
+            RefName::LocalBranch(branch) => format!("branch: {branch}"),
+        };
+
+        let import_status = match (remote_ref.target.is_absent(), ref_target.is_present()) {
+            (true, true) => ImportStatus::New,
+            (false, false) => ImportStatus::Deleted,
+            _ => ImportStatus::Updated,
```
Comment (yuja) [inline cli/src/git_util.rs:247]: nit: I would make both use `is_absent()` (or `is_present()`) so the table shows the difference `(false, true)` and `(true, false)`.

File: cli/src/commands/git.rs [hunk 12]
```diff
@@ -537,7 +537,7 @@ fn cmd_git_fetch(
             GitFetchError::InternalGitError(err) => map_git_error(err),
             _ => user_error(err),
         })?;
-        print_git_import_stats(ui, &stats.import_stats)?;
+        print_git_import_stats(ui, &stats.import_stats, Some(tx.repo().base_repo()))?;
```
Comment (yuja) [inline cli/src/commands/git.rs:540]: Pass `tx.repo()`.

`tx.repo().base_repo()` doesn't know the change applied by `git::fetch()`.

File: cli/src/commands/git.rs [hunk 13]
```diff
@@ -726,7 +726,7 @@ fn do_git_clone(
         r#"Fetching into new repo in "{}""#,
         wc_path.display()
     )?;
-    let mut workspace_command = command.for_loaded_repo(ui, workspace, repo)?;
+    let mut workspace_command = command.for_loaded_repo(ui, workspace, repo.clone())?;
```
Comment (yuja) [inline cli/src/commands/git.rs:729]: nit: unrelated change

File: cli/src/git_util.rs (hunk 8 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs:183]: resolved

File: cli/src/git_util.rs (hunk 9 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs]: resolved

File: cli/src/git_util.rs (hunk 10 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs]: resolved

File: cli/src/git_util.rs (hunk 11 above)
Comment (0xdeafbeef) [reply cli/src/git_util.rs:240]: resolved

File: cli/src/commands/git.rs (hunk 12 above)
Comment (0xdeafbeef) [reply cli/src/commands/git.rs]: resolved

File: cli/src/commands/git.rs (hunk 7 above)
Comment (0xdeafbeef) [reply cli/src/commands/git.rs]: resolved

Comment (0xdeafbeef) [reply cli/src/commands/git.rs]: `jj git clone git@github.com:martinvonz/jj.git  --colocate` 
now gives
```
Fetching into new repo in "/tmp/tmp.nCGWca6Ajb/jj"
branch: CacheMyWorkingCopy@origin        [new] untracked
branch: arxanas/ignore@origin            [new] untracked
branch: aseipp/push-mupwvrwmxuvm@origin  [new] untracked
branch: aseipp/push-mwwotvxyruwp@origin  [new] untracked
branch: aseipp/push-oolmkoukrxxt@origin  [new] untracked
branch: aseipp/push-srtxpxttmqlr@origin  [new] untracked
branch: aseipp/push-tuxwuxlpsttr@origin  [new] untracked
branch: aseipp/push-uytvkxyqyspn@origin  [new] untracked
branch: aseipp/push-zkyqxypkvulk@origin  [new] untracked
branch: gh-pages@origin                  [new] untracked
branch: ig/contributing@origin           [new] untracked
branch: ig/faq-move@origin               [new] untracked
branch: ig/retrytest@origin              [new] untracked
branch: ig/tmp-retrytest-with-testfail@origin [new] untracked
branch: jt/ignore@origin                 [new] untracked
branch: main@origin                      [new] untracked
branch: missingop@origin                 [new] untracked
branch: necauqua/signing@origin          [new] untracked
branch: pager@origin                     [new] untracked
branch: prev-next@origin                 [new] untracked
branch: push-5640745a446c47efb1bd819dd4c95a66@origin [new] untracked
branch: push-numyszkprmty@origin         [new] untracked
branch: push-onymwnywputl@origin         [new] untracked
branch: push-ppomxyyutuqm@origin         [new] untracked
branch: push-pvwusmnnpwul@origin         [new] untracked
branch: push-qvrktsnwpkzu@origin         [new] untracked
branch: push-rmzwokuwmwtx@origin         [new] untracked
branch: push-ruzwxounnsut@origin         [new] untracked
branch: push-rvuprqxytrtu@origin         [new] untracked
branch: push-skzwxzopxszv@origin         [new] untracked
branch: push-slnqmxmxyuvr@origin         [new] untracked
branch: push-sovruzpmulyr@origin         [new] untracked
branch: push-sxvnsknzvxku@origin         [new] untracked
branch: push-tlpqtmtonpqp@origin         [new] untracked
branch: push-ytvunlzsxzvx@origin         [new] untracked
branch: push-zmyoqzynpltt@origin         [new] untracked
branch: talp/moqlklnlrywq@origin         [new] untracked
branch: talp/oqpyxkqkuyvm@origin         [new] untracked
branch: testing@origin                   [new] untracked
tag: v0.10.0                             [new]
tag: v0.11.0                             [new]
tag: v0.12.0                             [new]
tag: v0.13.0                             [new]
tag: v0.14.0                             [new]
tag: v0.3.0                              [new]
tag: v0.3.1                              [new]
tag: v0.3.2                              [new]
tag: v0.3.3                              [new]
tag: v0.4.0                              [new]
tag: v0.5.0                              [new]
tag: v0.5.1                              [new]
tag: v0.6.0                              [new]
tag: v0.6.1                              [new]
tag: v0.7.0                              [new]
tag: v0.8.0                              [new]
tag: v0.9.0                              [new]
Working copy now at: uwlkpvyw c8c91829 (empty) (no description set)
Parent commit      : npnmltou 89cf6a00 main | graphlog: refactor graphlog::Edge enum and its usage
Added 334 files, modified 0 files, removed 0 files
```

**Comment by 0xdeafbeef** [comment]
I'm not sure what to do with reference names that are too long. For now, padding is limited to 40 characters; perhaps we should consider using some kind of ellipsis

**Comment by martinvonz** [comment]
> I'm not sure what to do with reference names that are too long. For now, padding is limited to 40 characters; perhaps we should consider using some kind of ellipsis

Seems fine for now. Padding to accommodate the longer name also seems fine.

**Comment by 0xdeafbeef** [comment]
I've added color for branch/tag name, alignment if both tags and branches are present.
image

File: cli/src/commands/git.rs [hunk 14]
```diff
@@ -447,7 +447,7 @@ fn init_git_refs(
     if !tx.mut_repo().has_changes() {
         return Ok(repo);
     }
-    print_git_import_stats(ui, &stats)?;
+    print_git_import_stats(ui, repo.as_ref(), &stats, false)?;
```
Comment (yuja) [inline cli/src/commands/git.rs:450]: nit: `tx.repo()`

File: cli/src/commands/git.rs (hunk 13 above)
Comment (yuja) [inline cli/src/commands/git.rs:729]: nit: remove this `.clone()`

File: cli/src/commands/git.rs [hunk 15]
```diff
@@ -751,7 +751,7 @@ fn do_git_clone(
             unreachable!("we didn't provide any globs")
         }
     })?;
-    print_git_import_stats(ui, &stats.import_stats)?;
+    print_git_import_stats(ui, repo.as_ref(), &stats.import_stats, true)?;
```
Comment (yuja) [inline cli/src/commands/git.rs:754]: `fetch_tx.repo()`

File: cli/src/git_util.rs [hunk 16]
```diff
@@ -172,17 +176,140 @@ pub fn with_remote_git_callbacks<T>(
...
+                if repo.view().get_remote_branch(branch, remote).is_tracking() {
+                    tracking_status = TrackingStatus::Tracked;
+                }
+                (format!("{branch}@{remote}"), RefKind::Branch)
+            }
+            RefName::Tag(tag) => {
+                tracking_status = TrackingStatus::NotApplicable;
+                (tag.clone(), RefKind::Tag)
+            }
+            RefName::LocalBranch(branch) => (branch.clone(), RefKind::Branch),
```
Comment (yuja) [inline cli/src/git_util.rs:247]: nit: set/return `(ref_name, ref_kind, tracking_status)` by `LocalBranch` consistently?

Under the hood, `Tag` and `LocalBranch` (which is `@git` branch) are always `Tracked`. We can of course map these to `NotApplicable` to suppress redundant output.

File: cli/src/git_util.rs [hunk 17]
```diff
@@ -172,17 +176,140 @@ pub fn with_remote_git_callbacks<T>(
...
+        };
+
+        let import_status = match self.import_status {
+            ImportStatus::New => "new",
+            ImportStatus::Deleted => "deleted",
+            ImportStatus::Updated => "updated",
+        };
+
+        // This ensures alignment of the statuses that follow
+        let padded_ref_name = format!("{:width$}", self.ref_name, width = max_ref_name_width);
```
Comment (yuja) [inline cli/src/git_util.rs:283]: we'll have to do something like `"{ref_name}{pad:pad_width}", pad = "", pad_width = saturating_sub(max_ref_name_width, ref_name.width())` (untested).

Otherwise the width of `self.ref_name` would be calculated by using `.len()`.

File: cli/src/commands/git.rs (hunk 7 above)
Comment (yuja) [reply cli/src/commands/git.rs]: Oops, this looks verbose than I expected. Maybe better for `jj git clone` to not show ref changes? What do you think?

**Review by yuja — APPROVED** [review]
Looks mostly good, thanks!

File: CHANGELOG.md [hunk 18]
```diff
@@ -45,6 +45,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#1252](https://github.com/martinvonz/jj/issues/1252),
   [#2971](https://github.com/martinvonz/jj/issues/2971)). This may become the
   default depending on feedback.
+* `jj git fetch` now automatically prints new remote branches and tags by default.
```
Comment (yuja) [inline CHANGELOG.md:48]: nit: insert blank line like the other entries
